extensions/ql-vscode/src/data-extensions-editor/extension-pack-picker.ts

```diff
@@ -303,12 +308,145 @@ async function pickNewExtensionPack(
   }
 
   const name = matches.groups.name;
-  const packPath = join(workspaceFolder.path, name);
+  const packPath = join(workspaceFolder.uri.fsPath, name);
 
   if (await pathExists(packPath)) {
     return undefined;
   }
 
+  return writeExtensionPack(packPath, packName, databaseItem.language);
+}
+
+async function autoCreateExtensionPack(
+  name: string,
+  language: string,
+  extensionPacksInfo: QlpacksInfo,
+  logger: NotificationLogger,
+): Promise<ExtensionPack | undefined> {
+  const workspaceFolder = await autoPickWorkspaceFolder(language);
+  if (!workspaceFolder) {
+    return undefined;
+  }
+
+  const packName = autoNameExtensionPack(name, language);
+  if (!packName) {
+    void showAndLogErrorMessage(
+      logger,
+      `Could not automatically name extension pack for database ${name}`,
+    );
+
+    return undefined;
+  }
+
+  const existingExtensionPackPaths = extensionPacksInfo[packName];
+  if (existingExtensionPackPaths?.length === 1) {
+    let extensionPack: ExtensionPack;
+    try {
+      extensionPack = await readExtensionPack(existingExtensionPackPaths[0]);
+    } catch (e: unknown) {
+      void showAndLogErrorMessage(
+        logger,
+        `Could not read extension pack ${packName}`,
+        {
+          fullMessage: `Could not read extension pack ${packName} at ${
+            existingExtensionPackPaths[0]
+          }: ${getErrorMessage(e)}`,
+        },
+      );
+
+      return undefined;
+    }
+
+    return extensionPack;
+  } else if (existingExtensionPackPaths?.length > 1) {
+    void showAndLogErrorMessage(
+      logger,
+      `Extension pack ${packName} resolves to multiple paths`,
+      {
+        fullMessage: `Extension pack ${packName} resolves to multiple paths: ${existingExtensionPackPaths.join(
+          ", ",
+        )}`,
+      },
+    );
+
+    return undefined;
+  }
+
+  const matches = packNameRegex.exec(packName);
+  if (!matches?.groups) {
+    void showAndLogErrorMessage(
+      logger,
+      `Extension pack ${packName} does not have a valid name`,
+    );
+
+    return undefined;
+  }
+
+  const unscopedName = matches.groups.name;
+  const packPath = join(workspaceFolder.uri.fsPath, unscopedName);
+
+  if (await pathExists(packPath)) {
+    void showAndLogErrorMessage(
+      logger,
+      `Directory ${packPath} already exists for extension pack ${packName}`,
+    );
+
+    return undefined;
+  }
+
+  return writeExtensionPack(packPath, packName, language);
+}
+
+async function autoPickWorkspaceFolder(
+  language: string,
+): Promise<WorkspaceFolder | undefined> {
+  const workspaceFolders = getOnDiskWorkspaceFoldersObjects();
+
+  if (workspaceFolders.length === 1) {
+    return workspaceFolders[0];
+  }
+  const starterWorkspaceFolderForLanguage = workspaceFolders.find(
+    (folder) => folder.name === `codeql-custom-queries-${language}`,
+  );
+  if (starterWorkspaceFolderForLanguage) {
+    return starterWorkspaceFolderForLanguage;
```

Comment:
Should we check for the presence of `queries.xml` with the right language?

Reply:
I don't think we need to. The extension pack doesn't actually depend on the folder where it's placed, we just need some place to put it. It'll still work if an extension pack for Java is placed in the `codeql-custom-queries-csharp` workspace folder.